### PR TITLE
Refactor `ferrocene/ci/split-tasks.py`

### PR DESCRIPTION
**Pull request by Urhengulas**
This PR
- adds type hints to most functions
- refactors `find_all_compiletests` a bit

**Question:** `find_all_compiletests` is written rather generically, but I only see it called one time in the whole repository? Why is it written like that? If it does not need to be that generic we can also change the types of the arguments to be not `None`.

### CODE REVIEW

File: ferrocene/ci/split-tasks.py [hunk 1]
```diff
@@ -18,29 +19,27 @@
 REPOSITORY_ROOT = pathlib.Path(__file__).resolve().parent.parent.parent
 
 
-def find_all_compiletests(*, exclude=None, deprioritize=None):
+def find_all_compiletests(exclude: list[str] | None, deprioritize: list[str] | None):
     if exclude is None:
         exclude = []
     if deprioritize is None:
-        prioritize = []
+        deprioritize = []
```
Comment (Hoverbear) [inline ferrocene/ci/split-tasks.py:26]: Oops 😅

**Review by Hoverbear — APPROVED** [review]

File: ferrocene/ci/split-tasks.py (hunk 1 above)
Comment (Urhengulas) [reply ferrocene/ci/split-tasks.py:26]: Happens 😉

**Comment by Urhengulas** [comment]
bors try

**Comment by Urhengulas** [comment]
bors merge

**Comment by Urhengulas** [comment]
The error is:

```
  File "/home/ci/project/ferrocene/ci/split-tasks.py", line 52
    type JobsDefinition = dict[str, Kind]
         ^^^^^^^^^^^^^^
SyntaxError: invalid syntax
```
Link to logs: https://app.circleci.com/pipelines/github/ferrocene/ferrocene/2809/workflows/a3e73e72-6feb-4553-858f-eb4aff03cc56/jobs/13282?invite=true#step-111-1741_62

> The [type](https://docs.python.org/3/reference/simple_stmts.html#type) statement is new in Python 3.12

So I assume our CI uses an older python version?

**Comment by Urhengulas** [comment]
bors try

**Review by Hoverbear — APPROVED** [review]

**Comment by Urhengulas** [comment]
bors merge

**Pull request merged**